File: src/main/java/com/demo/viettelpost/ViettelPostService.java
package com.demo.viettelpost;

import com.demo.viettelpost.ViettelPostDTO.*;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.stereotype.Service;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Service for interacting with Viettel Post API
 */
@Service
public class ViettelPostService {

    private static final String API_BASE = "https://partner.viettelpost.vn/v2";
    
    // Pre-configured token (your token)
    private static final String DEFAULT_TOKEN = "41A7F86612D4357C125529D1878BBE38";
    
    // Token storage (in-memory, can be replaced with database)
    private final Map<String, ViettelPostToken> tokenStorage = new ConcurrentHashMap<>();
    
    private final HttpClient httpClient;
    private final ObjectMapper objectMapper;
    
    // Service code to name mapping
    private static final Map<String, String> SERVICE_NAMES = Map.of(
        "VCN", "Chuyển phát nhanh",
        "VCBO", "Chuyển phát nhanh theo bộ",
        "VHT", "Phát hỏa tốc",
        "PTN", "Phát trong ngày",
        "PHS", "Phát hẹn giờ",
        "VBS", "Chuyển phát tiêu chuẩn",
        "SCOD", "Dịch vụ thu hộ COD",
        "VTK", "Tiết kiệm",
        "VBE", "Chuyển phát tiết kiệm"
    );

    public ViettelPostService() {
        this.httpClient = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(30))
            .build();
        this.objectMapper = new ObjectMapper();
        
        // Store default token
        ViettelPostToken defaultToken = new ViettelPostToken("default", DEFAULT_TOKEN, "viettelpost_user");
        tokenStorage.put("default", defaultToken);
    }

    /**
     * Get default/stored token
     */
    public String getDefaultToken() {
        return DEFAULT_TOKEN;
    }

    /**
     * Store a token
     */
    public void storeToken(String userId, String token, String username) {
        ViettelPostToken vptToken = new ViettelPostToken(userId, token, username);
        tokenStorage.put(userId, vptToken);
    }

    /**
     * Get token by user ID
     */
    public ViettelPostToken getToken(String userId) {
        ViettelPostToken token = tokenStorage.get(userId);
        if (token != null && !token.isExpired()) {
            token.updateLastUsed();
            return token;
        }
        return null;
    }

    /**
     * Check if token is valid
     */
    public boolean hasValidToken(String userId) {
        ViettelPostToken token = tokenStorage.get(userId);
        return token != null && !token.isExpired();
    }

    /**
     * Login to Viettel Post API
     */
    public ViettelPostToken login(String username, String password) throws Exception {
        // Check if already has valid token
        if (hasValidToken(username)) {
            return getToken(username);
        }

        // Step 1: Login to get temporary token
        Map<String, String> loginBody = Map.of("USERNAME", username, "PASSWORD", password);
        String loginJson = objectMapper.writeValueAsString(loginBody);

        HttpRequest loginRequest = HttpRequest.newBuilder()
            .uri(URI.create(API_BASE + "/user/Login"))
            .header("Content-Type", "application/json")
            .POST(HttpRequest.BodyPublishers.ofString(loginJson))
            .build();

        HttpResponse<String> loginResponse = httpClient.send(loginRequest, HttpResponse.BodyHandlers.ofString());
        JsonNode loginData = objectMapper.readTree(loginResponse.body());

        if (loginData.get("status").asInt() != 200 || loginData.get("data") == null) {
            throw new RuntimeException("Invalid credentials");
        }

        String tempToken = loginData.get("data").get("token").asText();

        // Step 2: Get long-term token
        HttpRequest connectRequest = HttpRequest.newBuilder()
            .uri(URI.create(API_BASE + "/user/ownerconnect"))
            .header("Content-Type", "application/json")
            .header("Token", tempToken)
            .POST(HttpRequest.BodyPublishers.ofString("{}"))
            .build();

        HttpResponse<String> connectResponse = httpClient.send(connectRequest, HttpResponse.BodyHandlers.ofString());
        JsonNode connectData = objectMapper.readTree(connectResponse.body());

        if (connectData.get("status").asInt() != 200 || connectData.get("data") == null) {
            throw new RuntimeException("Failed to get long-term token");
        }

        String longTermToken = connectData.get("data").get("token").asText();

        // Store token
        ViettelPostToken vptToken = new ViettelPostToken(username, longTermToken, username);
        tokenStorage.put(username, vptToken);

        return vptToken;
    }

    /**
     * Get all provinces
     */
    public List<Province> getProvinces(String token) throws Exception {
        HttpRequest request = HttpRequest.newBuilder()
            .uri(URI.create(API_BASE + "/categories/listProvince"))
            .header("Token", token)
            .GET()
            .build();

        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        JsonNode data = objectMapper.readTree(response.body());

        if (data.get("status").asInt() == 200 && data.has("data")) {
            return objectMapper.readValue(
                data.get("data").toString(),
                new TypeReference<List<Province>>() {}
            );
        }
        return Collections.emptyList();
    }

    /**
     * Get districts by province ID
     */
    public List<District> getDistricts(String token, int provinceId) throws Exception {
        HttpRequest request = HttpRequest.newBuilder()
            .uri(URI.create(API_BASE + "/categories/listDistrict?provinceId=" + provinceId))
            .header("Token", token)
            .GET()
            .build();

        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        JsonNode data = objectMapper.readTree(response.body());

        if (data.get("status").asInt() == 200 && data.has("data")) {
            return objectMapper.readValue(
                data.get("data").toString(),
                new TypeReference<List<District>>() {}
            );
        }
        return Collections.emptyList();
    }

    /**
     * Get wards by district ID
     */
    public List<Ward> getWards(String token, int districtId) throws Exception {
        HttpRequest request = HttpRequest.newBuilder()
            .uri(URI.create(API_BASE + "/categories/listWards?districtId=" + districtId))
            .header("Token", token)
            .GET()
            .build();

        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        JsonNode data = objectMapper.readTree(response.body());

        if (data.get("status").asInt() == 200 && data.has("data")) {
            return objectMapper.readValue(
                data.get("data").toString(),
                new TypeReference<List<Ward>>() {}
            );
        }
        return Collections.emptyList();
    }

    /**
     * Get complete address
     */
    public CompleteAddress getCompleteAddress(String token, int provinceId, Integer districtId, Integer wardId) throws Exception {
        CompleteAddress address = new CompleteAddress();
        
        List<Province> provinces = getProvinces(token);
        Province province = provinces.stream()
            .filter(p -> p.getProvinceId() == provinceId)
            .findFirst()
            .orElse(null);
        address.setProvince(province);

        if (districtId != null) {
            List<District> districts = getDistricts(token, provinceId);
            District district = districts.stream()
                .filter(d -> d.getDistrictId() == districtId)
                .findFirst()
                .orElse(null);
            address.setDistrict(district);

            if (wardId != null && district != null) {
                List<Ward> wards = getWards(token, districtId);
                Ward ward = wards.stream()
                    .filter(w -> w.getWardsId() == wardId)
                    .findFirst()
                    .orElse(null);
                address.setWard(ward);
            }
        }

        // Build full address string
        StringBuilder fullAddr = new StringBuilder();
        if (address.getWard() != null) {
            fullAddr.append(address.getWard().getWardsName()).append(", ");
        }
        if (address.getDistrict() != null) {
            fullAddr.append(address.getDistrict().getDistrictName()).append(", ");
        }
        if (address.getProvince() != null) {
            fullAddr.append(address.getProvince().getProvinceName());
        }
        address.setFullAddress(fullAddr.toString());

        return address;
    }

    /**
     * Calculate price for a single service
     */
    public PriceData calculatePrice(String token, PriceRequest request) throws Exception {
        String requestJson = objectMapper.writeValueAsString(request);

        HttpRequest httpRequest = HttpRequest.newBuilder()
            .uri(URI.create(API_BASE + "/order/getPrice"))
            .header("Content-Type", "application/json")
            .header("Token", token)
            .POST(HttpRequest.BodyPublishers.ofString(requestJson))
            .build();

        HttpResponse<String> response = httpClient.send(httpRequest, HttpResponse.BodyHandlers.ofString());
        JsonNode data = objectMapper.readTree(response.body());

        if (data.get("status").asInt() == 200 && data.has("data")) {
            return objectMapper.readValue(data.get("data").toString(), PriceData.class);
        }
        return null;
    }

    /**
     * Get all available shipment options
     */
    public List<ShipmentOption> getShipmentOptions(String token, PriceRequest baseRequest) throws Exception {
        List<ShipmentOption> options = new ArrayList<>();
        String[] serviceCodes = {"VCN", "VCBO", "VHT", "PTN", "PHS", "VBS", "SCOD"};

        for (String serviceCode : serviceCodes) {
            try {
                PriceRequest request = new PriceRequest();
                request.setProductWeight(baseRequest.getProductWeight());
                request.setProductPrice(baseRequest.getProductPrice());
                request.setMoneyCollection(baseRequest.getMoneyCollection());
                request.setSenderProvince(baseRequest.getSenderProvince());
                request.setSenderDistrict(baseRequest.getSenderDistrict());
                request.setReceiverProvince(baseRequest.getReceiverProvince());
                request.setReceiverDistrict(baseRequest.getReceiverDistrict());
                request.setProductType(baseRequest.getProductType());
                request.setNationalType(baseRequest.getNationalType());
                request.setOrderService(serviceCode);

                PriceData priceData = calculatePrice(token, request);
                if (priceData != null && priceData.getMoneyTotal() > 0) {
                    String serviceName = SERVICE_NAMES.getOrDefault(serviceCode, serviceCode);
                    options.add(new ShipmentOption(serviceCode, serviceName, priceData));
                }
            } catch (Exception e) {
                // Service not available for this route, skip
            }
        }

        return options;
    }

    /**
     * Get service name from code
     */
    public String getServiceName(String code) {
        return SERVICE_NAMES.getOrDefault(code, code);
    }

    /**
     * Get all stored tokens (for admin)
     */
    public Map<String, ViettelPostToken> getAllTokens() {
        return new HashMap<>(tokenStorage);
    }

    /**
     * Remove token
     */
    public void removeToken(String userId) {
        tokenStorage.remove(userId);
    }

    /**
     * Clean expired tokens
     */
    public int cleanExpiredTokens() {
        int cleaned = 0;
        Iterator<Map.Entry<String, ViettelPostToken>> it = tokenStorage.entrySet().iterator();
        while (it.hasNext()) {
            if (it.next().getValue().isExpired()) {
                it.remove();
                cleaned++;
            }
        }
        return cleaned;
    }
}

File: viettelpost/viettelPostService.js
// Viettel Post Service - Business logic for Viettel Post API
const axios = require('axios');

const API_BASE = 'https://partner.viettelpost.vn/v2';

class ViettelPostService {
    constructor() {
        this.apiBase = API_BASE;
    }

    // Login and get tokens
    async login(username, password) {
        try {
            // Step 1: Login to get temporary token
            const loginResponse = await axios.post(`${this.apiBase}/user/Login`, {
                USERNAME: username,
                PASSWORD: password
            }, {
                headers: {
                    'Content-Type': 'application/json'
                }
            });

            if (loginResponse.data.status !== 200 || !loginResponse.data.data?.token) {
                throw new Error('Invalid credentials');
            }

            const tempToken = loginResponse.data.data.token;

            // Step 2: Get long-term token
            const connectResponse = await axios.post(`${this.apiBase}/user/ownerconnect`, {}, {
                headers: {
                    'Content-Type': 'application/json',
                    'Token': tempToken
                }
            });

            if (connectResponse.data.status !== 200 || !connectResponse.data.data?.token) {
                throw new Error('Failed to get long-term token');
            }

            const longTermToken = connectResponse.data.data.token;
            const userData = connectResponse.data.data;

            return {
                token: longTermToken,
                tempToken: tempToken,
                userId: userData.userId || username,
                username: username,
                fullData: userData
            };
        } catch (error) {
            console.error('Login error:', error.response?.data || error.message);
            throw new Error(error.response?.data?.message || error.message || 'Login failed');
        }
    }

    // Get all provinces
    async getProvinces(token) {
        try {
            const response = await axios.get(`${this.apiBase}/categories/listProvince`, {
                headers: { 'Token': token }
            });
            return response.data;
        } catch (error) {
            throw new Error('Failed to fetch provinces: ' + error.message);
        }
    }

    // Get districts by province ID
    async getDistricts(token, provinceId) {
        try {
            const response = await axios.get(`${this.apiBase}/categories/listDistrict?provinceId=${provinceId}`, {
                headers: { 'Token': token }
            });
            return response.data;
        } catch (error) {
            throw new Error('Failed to fetch districts: ' + error.message);
        }
    }

    // Get wards by district ID
    async getWards(token, districtId) {
        try {
            const response = await axios.get(`${this.apiBase}/categories/listWards?districtId=${districtId}`, {
                headers: { 'Token': token }
            });
            return response.data;
        } catch (error) {
            throw new Error('Failed to fetch wards: ' + error.message);
        }
    }

    // Get complete address hierarchy
    async getCompleteAddress(token, provinceId, districtId, wardId) {
        try {
            const [provinces, districts, wards] = await Promise.all([
                this.getProvinces(token),
                districtId ? this.getDistricts(token, provinceId) : Promise.resolve({ data: [] }),
                wardId ? this.getWards(token, districtId) : Promise.resolve({ data: [] })
            ]);

            const province = provinces.data?.find(p => p.PROVINCE_ID == provinceId);
            const district = districts.data?.find(d => d.DISTRICT_ID == districtId);
            const ward = wards.data?.find(w => w.WARDS_ID == wardId);

            return {
                province: province || null,
                district: district || null,
                ward: ward || null,
                fullAddress: [ward?.WARDS_NAME, district?.DISTRICT_NAME, province?.PROVINCE_NAME]
                    .filter(Boolean)
                    .join(', ')
            };
        } catch (error) {
            throw new Error('Failed to fetch complete address: ' + error.message);
        }
    }

    // Calculate shipping price for a single service
    async calculatePrice(token, priceData) {
        try {
            const response = await axios.post(`${this.apiBase}/order/getPrice`, priceData, {
                headers: {
                    'Content-Type': 'application/json',
                    'Token': token
                }
            });
            return response.data;
        } catch (error) {
            throw new Error('Failed to calculate price: ' + error.message);
        }
    }

    // Calculate prices for all available services
    async calculateAllPrices(token, priceData) {
        try {
            const response = await axios.post(`${this.apiBase}/order/getPriceAll`, priceData, {
                headers: {
                    'Content-Type': 'application/json',
                    'Token': token
                }
            });
            return response.data;
        } catch (error) {
            throw new Error('Failed to calculate all prices: ' + error.message);
        }
    }

    // Get all shipment options with enhanced data
    async getShipmentOptions(token, shipmentData) {
        try {
            // Try to get all prices first
            const priceResponse = await this.calculateAllPrices(token, shipmentData);
            
            if (priceResponse.status === 200 && priceResponse.data) {
                // Enhance shipment options with additional info
                const services = Array.isArray(priceResponse.data) ? priceResponse.data : [priceResponse.data];
                
                return services.map(service => ({
                    serviceCode: service.MA_DV_CHINH,
                    serviceName: service.TEN_DICHVU || service.MA_DV_CHINH,
                    baseFee: service.GIA_CUOC || 0,
                    extraServicesFee: service.GIA_CUOC_DV_CON || 0,
                    insuranceFee: service.PHI_BAO_HIEM || 0,
                    remoteFee: service.PHI_XA_TRAM || 0,
                    totalFee: service.TONG_CUOC || service.MONEY_TOTAL || 0,
                    deliveryTime: service.THOI_GIAN || 'N/A',
                    rawData: service
                }));
            }

            // If getPriceAll doesn't work, try individual services
            const services = ['VCN', 'VCBO', 'VHT', 'PTN', 'PHS', 'VBS', 'SCOD'];
            const results = [];

            for (const serviceCode of services) {
                try {
                    const singlePrice = await this.calculatePrice(token, {
                        ...shipmentData,
                        ORDER_SERVICE: serviceCode
                    });
                    
                    if (singlePrice.status === 200 && singlePrice.data) {
                        results.push({
                            serviceCode: serviceCode,
                            serviceName: this.getServiceName(serviceCode),
                            baseFee: singlePrice.data.MONEY_TOTAL_FEE || 0,
                            vatFee: singlePrice.data.MONEY_VAT || 0,
                            collectionFee: singlePrice.data.MONEY_COLLECTION_FEE || 0,
                            otherFee: singlePrice.data.MONEY_OTHER_FEE || 0,
                            totalFee: singlePrice.data.MONEY_TOTAL || 0,
                            deliveryTime: singlePrice.data.KPI_HT ? `${singlePrice.data.KPI_HT} giờ` : 'N/A',
                            rawData: singlePrice.data
                        });
                    }
                } catch (e) {
                    // Service not available for this route, skip
                }
            }

            if (results.length === 0) {
                throw new Error('No shipment options available for this route');
            }

            return results;
        } catch (error) {
            throw new Error('Failed to get shipment options: ' + error.message);
        }
    }

    // Get service name from code
    getServiceName(code) {
        const serviceNames = {
            'VCN': 'Chuyển phát nhanh',
            'VCBO': 'Chuyển phát nhanh theo bộ',
            'VHT': 'Phát hỏa tốc',
            'PTN': 'Phát trong ngày',
            'PHS': 'Phát hẹn giờ',
            'VBS': 'Chuyển phát tiêu chuẩn',
            'SCOD': 'Dịch vụ thu hộ COD',
            'LCOD': 'COD Lite',
            'VTK': 'Tiết kiệm',
            'VBE': 'Chuyển phát tiết kiệm',
            'PHT': 'Phát hàng thu tiền'
        };
        return serviceNames[code] || code;
    }

    // Create order (if you want to extend functionality)
    async createOrder(token, orderData) {
        try {
            const response = await axios.post(`${this.apiBase}/order/createOrder`, orderData, {
                headers: {
                    'Content-Type': 'application/json',
                    'Token': token
                }
            });
            return response.data;
        } catch (error) {
            throw new Error('Failed to create order: ' + error.message);
        }
    }

    // Get order info
    async getOrderInfo(token, orderNumber) {
        try {
            const response = await axios.get(`${this.apiBase}/order/getOrderInfoByCode?orderNumber=${orderNumber}`, {
                headers: { 'Token': token }
            });
            return response.data;
        } catch (error) {
            throw new Error('Failed to get order info: ' + error.message);
        }
    }
}

module.exports = new ViettelPostService();
